### PR TITLE
Refactor project structure and improve agent calculation

### PR DESCRIPTION
**Pull request by xoxys**
First, thank you for starting an autoscale implementation for Woodpecker. For me, this was an obstacle to finally get away from Drone. 

When I tried the autoscaler on my Woodpecker setup, I had some problems with the calculation. Due to the short interval (1s) and the missing `amountActiveAgentsFromPool` in the calculation, the autoscaler always created all agents up to the maximum, because in the calculation `diffAmountAgents := math.Ceil(neededAmountAgents - availableAgents)` only `availableAgents` is considered. But as new servers need some time to be provisioned, they do not report their capacity in the meantime and are therefore not included in `availableAgents` on the next reconciliation run. As a result, the autoscaler is provisioning another one.

This behavior was fixed by considering available Agents as well as agents from the Pool. This is still not perfect, if "something" happens during the server provisioning that prevents connection to the woodpecker server, the autoscaler will not create a new agent to fulfill the workload requirements.

As a long term solution, I think using a proper async task runner as proposed in https://github.com/woodpecker-ci/autoscaler/issues/2 could help to improve the reconciliation process. Tasks will be tracked, monitored and health-checked and failing tasks can be retried or rolled back.

I know this PR is a big one, that is mixing new features, fixes and refactoring but it somehow every change depends on the other and I didn't find a proper way to split it into smaller pieces.

### CODE REVIEW

**Comment by 6543** [comment]
did made an exeption from squash-merge to keep the commit history in this case

**Review by 6543 — COMMENTED** [review]
ah LGTM by the way :D

**Comment by xoxys** [comment]
Thanks for the fast review!

File: .golangci.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    # gin force us to use string as context key
+    - path: server/store/context.go
+      linters:
+        - staticcheck
+        - revive
+
+    # let cli use print and panic
+    - path: "cmd/*|cli/*"
+      linters:
+        - forbidigo
```
Comment (anbraten) [inline .golangci.yml:52]: Could you add the ci job as well please

File: cmd/woodpecker-autoscaler/flags.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	},
+	&cli.BoolFlag{
+		Name:    "grpc-secure",
+		Value:   false,
+		Usage:   "use secure grpc connection to the woodpecker server",
+		EnvVars: []string{"WOODPECKER_GRPC_SECURE"},
+	},
+	&cli.StringFlag{
+		Name:    "provider",
+		Value:   "hetznercloud",
```
[the remark this answers is not in the document]
Comment (xoxys) [reply cmd/woodpecker-autoscaler/flags.go:73]: Why should we remove the provider default?

File: .golangci.yml (hunk 1 above)
Comment (xoxys) [reply .golangci.yml:52]: Sure :+1:

File: cmd/woodpecker-autoscaler/main.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	if err != nil {
+		return err
+	}
+
+	autoscaler := engine.NewAutoscaler(provider, client, config)
+
+	interval, err := time.ParseDuration(ctx.String("interval"))
+	if err != nil {
+		log.Error().Err(err).Msgf("cant parse reconciliation interval, use default: %v", optionIntervalDefault)
+		interval, _ = time.ParseDuration(optionIntervalDefault)
```
[the remark this answers is not in the document]
Comment (xoxys) [reply cmd/woodpecker-autoscaler/main.go:68]: Why not automatically fallback?

File: cmd/woodpecker-autoscaler/main.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+		Name:  "autoscaler",
+		Usage: "scale to the moon and back",
+		Flags: flags,
+		Before: func(ctx *cli.Context) error {
+			zerolog.SetGlobalLevel(zerolog.InfoLevel)
+			if ctx.IsSet("log-level") {
+				logLevelFlag := ctx.String("log-level")
+				lvl, err := zerolog.ParseLevel(logLevelFlag)
+				if err != nil {
+					log.Warn().Str("level", logLevelFlag).Msg("unknown logging level")
```
[the remark this answers is not in the document]
Comment (xoxys) [reply cmd/woodpecker-autoscaler/main.go]: Same question here, the return at this point is not set the default/fallback log level from line "96". It is the same behavior as on the server.

File: .golangci.yml (hunk 1 above)
Comment (xoxys) [reply .golangci.yml:52]: https://github.com/woodpecker-ci/autoscaler/pull/6

**Comment by anbraten** [comment]
🎉 This PR is included in version 0.1.0 🎉

The release is now available [here](https://github.com/woodpecker-ci/autoscaler/releases/tag/0.1.0)

Thank you for your contribution. ❤️📦🚀

**Comment by anbraten** [comment]
🎉 This PR is included in version 0.1.0 🎉

The release is now available [here](https://github.com/woodpecker-ci/autoscaler/releases/tag/0.1.0)

Thank you for your contribution. ❤️📦🚀

**Comment by anbraten** [comment]
🎉 This PR is included in version 0.1.0 🎉

The release is now available [here](https://github.com/woodpecker-ci/autoscaler/releases/tag/0.1.0)

Thank you for your contribution. ❤️📦🚀

**Comment by anbraten** [comment]
🎉 This PR is included in version 0.1.0 🎉

The release is now available [here](https://github.com/woodpecker-ci/autoscaler/releases/tag/0.1.0)

Thank you for your contribution. ❤️📦🚀

**Comment by anbraten** [comment]
🎉 This PR is included in version 0.1.0 🎉

The release is now available [here](https://github.com/woodpecker-ci/autoscaler/releases/tag/0.1.0)

Thank you for your contribution. ❤️📦🚀

**Comment by anbraten** [comment]
🎉 This PR is included in version 0.1.0 🎉

The release is now available [here](https://github.com/woodpecker-ci/autoscaler/releases/tag/0.1.0)

Thank you for your contribution. ❤️📦🚀

**Comment by anbraten** [comment]
🎉 This PR is included in version 0.1.0 🎉

The release is now available [here](https://github.com/woodpecker-ci/autoscaler/releases/tag/0.1.0)

Thank you for your contribution. ❤️📦🚀

**Pull request merged**